### PR TITLE
Alternate Steam Libraries

### PR DESCRIPTION
**Pull request by scfinniss**
I noticed while working on the Watcom C stuff that Chocolate Doom was not locating my IWADs.

What I've done here is scrape the alternative library locations from Steam's `libraryfolders.vdf` file and use them to add more search paths. The file itself is mostly key value pairs with a header. Here is the file from my setup for reference.

```
"LibraryFolders"
{
	"TimeNextStatsReport"		"1488133041"
	"ContentStatsID"		"198096677914456296"
	"1"		"D:\\Steam"
	"2"		"E:\\Steam"
	"3"		"C:\\Steam"
}
```

I specifically ignore `LibraryFolders`, thereafter I assume every second string is a path, as long as the preceding key is numeric. I make no real effort to understand the structure of the file beyond this.

It's important to note that the unlisted `C:\Program Files (x86)\Steam` is also a library. Presumably Steam always considers its install path a library.

I believe this may resolve issue #482 (excluding Mac and Linux support).

### CODE REVIEW

File: src/d_iwad.c [hunk 1]
```diff
@@ -345,11 +348,97 @@ static void CheckInstallRootPaths(void)
...
+    {
+        steam_libraries[num_steam_libraries] = M_StringReplace(dir, "\\\\",
+                                                               "\\");
+        ++num_steam_libraries;
+    }
+}
+
+static void CheckSteamLibrary(long library_file_size, char *library_file_buffer)
+{
+    long i, start;
```
Comment (fragglet) [inline src/d_iwad.c]: Please don't use type `long`, since it means very different things in different compilers. Please remove all uses of `long`.

File: src/d_iwad.c [hunk 2]
```diff
@@ -345,11 +348,97 @@ static void CheckInstallRootPaths(void)
...
+    boolean key, key_valid, value;
+    char* buffer;
```
Comment (fragglet) [inline src/d_iwad.c]: `char *buffer;`

File: src/d_iwad.c [hunk 3]
```diff
@@ -345,11 +348,97 @@ static void CheckInstallRootPaths(void)
...
+                    }
+                }
+
+                start = -1;
+            }
+        }
+        else if (start > -1 && key)
+        {
+            //Keys for libraries are always numeric
+            if (library_file_buffer[i] < '0' || library_file_buffer[i] > '9')
```
Comment (fragglet) [inline src/d_iwad.c]: Please use `isdigit()`.

File: src/d_iwad.c [hunk 4]
```diff
@@ -358,15 +447,53 @@ static void CheckSteamEdition(void)
...
+    AddSteamLibraryDir(install_path);
+
+    library_path = M_StringJoin(install_path, DIR_SEPARATOR_S,
+                                STEAM_LIBRARY_FOLDERS, NULL);
+
+    probe = M_FileCaseExists(library_path);
+
+    if (probe != NULL)
+    {
+        library_file = fopen(probe, "rb");
```
Comment (fragglet) [inline src/d_iwad.c]: Please factor this whole block out into a separate function.

File: src/d_iwad.c [hunk 5]
```diff
@@ -345,11 +348,97 @@ static void CheckInstallRootPaths(void)
...
+    boolean key, key_valid, value;
+    char* buffer;
+    start = -1;
+    key = true;
+    value = false;
+    key_valid = true;
+
+    for (i=0;i<library_file_size; ++i)
+    {
+        if (library_file_buffer[i] == '"')
```
Comment (fragglet) [inline src/d_iwad.c]: The logic in this function is very complicated and hard to follow. It looks like you're trying to do tokenization and higher-level logic in a single function. I think you can probably simplify this greatly by splitting out these into separate concerns.

**Comment by scfinniss** [comment]
The main change here is restructuring based on the feedback.
Changes in `CheckSteamEdition` have been mostly removed, replaced with one call to `CheckSteamLibraries`.

Then `CheckSteamLibrary` was replaced with
* `GetSteamNextLibPath`
* `IsValidSteamLibKey`
* `GetNextSteamLibToken`
This hopefully makes the logic easier to follow.

Instances of `long` replaced with `size_t`, hopefully I've not introduced any style issues this time.

**Comment by scfinniss** [comment]
I'm going to use the [reference from Valve](https://developer.valvesoftware.com/wiki/KeyValues#File_Format) to make some changes to how I parse the `libraryfolders.vdf`.

Is it worth moving this logic to a new file? (eg `m_steam.h`, `m_steam.c`)

File: src/d_iwad.c [hunk 6]
```diff
@@ -345,11 +348,198 @@ static void CheckInstallRootPaths(void)
...
+    }
+
+    library_file_size = M_FileLength(library_file);
+    library_file_buffer = malloc(library_file_size + 1);
+    count = fread(library_file_buffer, 1, library_file_size, library_file);
+    fclose(library_file);
+
+    if (library_file_size != count)
+    {
+        return NULL;
```
Comment (fragglet) [inline src/d_iwad.c]: free(library_file_buffer);

File: src/d_iwad.c [hunk 7]
```diff
@@ -345,11 +348,198 @@ static void CheckInstallRootPaths(void)
...
+                                //so we use this instead of NULL check
+    buffer = *lib_file_buffer;
+    buff_size = strlen(buffer);
+
+    for (i=0; i<buff_size; ++i)
+    {
+        if (buffer[i] == '"')
+        {
+            //first " is the start of the token
+            if (token == NULL)
```
Comment (fragglet) [inline src/d_iwad.c]: It becomes simpler if you just do two loops: one to find the starting ", another to find the ending ".

File: src/d_iwad.c [hunk 8]
```diff
@@ -345,11 +348,198 @@ static void CheckInstallRootPaths(void)
 
 // Check for Doom downloaded via Steam
 
+#define MAX_STEAM_LIBRARIES 16
+
+static char *steam_libraries[MAX_STEAM_LIBRARIES];
+static int num_steam_libraries = 0;
+
+static void AddSteamLibraryDir(char *dir)
```
Comment (fragglet) [inline src/d_iwad.c]: Do you need the `steam_libraries[]` array? Can't you just make this a wrapper around AddIWADDir?

File: src/d_iwad.c [hunk 9]
```diff
@@ -345,11 +348,198 @@ static void CheckInstallRootPaths(void)
...
+    for (;;)
+    {
+        token = GetNextSteamLibToken(lib_file_buffer);
+
+        if (token == NULL)
+        {
+            return NULL;
+        }
+
+        //file always starts with this line before key/value pairs
```
Comment (fragglet) [inline src/d_iwad.c]: So in that case, just put a call before the start of the loop in `CheckSteamLibraries()` that calls `GetNextSteamLibToken()` and checks it's equal to "LibraryFolders". Then this function should become much simpler.

File: src/d_iwad.c (hunk 8 above)
Comment (scfinniss) [reply src/d_iwad.c]: Yes, that array is note needed.

File: src/d_iwad.c (hunk 9 above)
Comment (scfinniss) [reply src/d_iwad.c]: I've changed how I parse the file. I treat the whole VDF file as key value pairs, and I don't bother checking that it starts with "LibraryFolders". The code is a lot simpler now.

File: src/d_iwad.c (hunk 7 above)
Comment (scfinniss) [reply src/d_iwad.c]: Thanks, this has helped.

**Comment by scfinniss** [comment]
I've addressed the reviews, and also made the tokenizer understand VDF more completely.

The tokenizer now handles tokens that are not always quoted, as well as ones with escaped characters.

To make the parsing simpler I've avoided the issue of subkeys completely. I ignore closing braces and treat the opening ones as part of a regular unquoted token.

So
```
"LibraryFolders"
{
	"TimeNextStatsReport"		"1505818818"
	"ContentStatsID"		"198096677914456296"
	"1"		"D:\\Steam"
	"2"		"E:\\Steam"
	"3"		"C:\\Steam"
}
```
Get tokenized to
```
"LibraryFolders"
{
"TimeNextStatsReport"
"1505818818"
"ContentStatsID"
"198096677914456296"
"1"
"D:\\Steam"
"2"
"E:\\Steam"
"3"
"C:\\Steam"
```
Every odd token is a key, every even token is a value.

File: src/d_iwad.c [hunk 10]
```diff
@@ -345,27 +348,245 @@ static void CheckInstallRootPaths(void)
...
+        {
+            continue;
+        }
+
+        quote = c == '"';
+        token = &buffer[i++];
+        break;
+    }
+
+    for (; i<buff_size; ++i)
```
Comment (fragglet) [inline src/d_iwad.c]: This ends up being complicated because you're trying to process different types of things using the same loop. You'd be better off just treating them separately. For example:

```
c = ... whatever
if (c == '"') {
    return ReadStringToken(.....);
} else if (c == '{') {
    return TOKEN_OPEN_BRACE;
} else if (c == '}') {
    return TOKEN_CLOSE_BRACE;
} else if (isalpha(c)) {
    return ReadKeyword(......);
} else {
    return TOKEN_UNKNOWN;
}

File: src/d_iwad.c [hunk 11]
```diff
@@ -345,27 +348,245 @@ static void CheckInstallRootPaths(void)
...
 
-    if (install_path == NULL)
+        AddIWADDir(subpath);
+    }
+}
+
+//Read whole of libraryfolders.vdf as string
+//silently fail is something goes wrong reading the file
+
+static char *GetSteamLibText(char *library_path)
```
[the remark this answers is not in the document]
Comment (scfinniss) [reply src/d_iwad.c]: Yes, I've got it working with `M_ReadFile()`.

The first difference I see between `M_ReadFile()` and my code is that I'm adding the `\0` to the end of the buffer. I assume I did that to ensure `strlen()` is guaranteed to work later. Not sure if that was ever necessary.

The other difference is that I'm using `malloc()`, where `M_ReadFile()` uses `Z_Malloc()`.

There are a few places where `malloc()` is used over `Z_Malloc()`, are these intentional or should `Z_Malloc()` always be used?

**Comment by mfrancis95** [comment]
@scfinniss any updates?

Otherwise I don't think it's good to keep a nearly 3 year-old PR up.

**Comment by scfinniss** [comment]
Real life got between me and finishing this, and I guess I forgot about it after awhile.

I'm not sure how much Chocolate Doom has changed since I worked on this, but I don't mind looking into it again if there's still interest in this feature?

**Comment by mfrancis95** [comment]
There have always been efforts to improve WAD-locating, even in recent times. I would say it's still worth it, but it just happens to be a fairly sizeable undertaking.
